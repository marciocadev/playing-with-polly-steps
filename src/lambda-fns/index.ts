import { Stream } from 'stream';
import { Engine, OutputFormat, PollyClient, SynthesizeSpeechCommand, SynthesizeSpeechInput, TextType } from '@aws-sdk/client-polly';
import { APIGatewayProxyEventV2, APIGatewayProxyResultV2 } from 'aws-lambda';

const pollyClient = new PollyClient({ region: process.env.AWS_REGION });

export const handler = async(event:APIGatewayProxyEventV2) => {

  if (!event?.body || !event?.pathParameters) {
    return {
      statusCode: 400,
    };
  }

  let { voice } = event.pathParameters;
  let text = event.body;

  const input:SynthesizeSpeechInput = {
    Text: text,
    VoiceId: voice,
    Engine: Engine.NEURAL,
    TextType: TextType.TEXT,
    OutputFormat: OutputFormat.MP3,
  };
  const command = new SynthesizeSpeechCommand(input);
  const synth = await pollyClient.send(command);

  const buffer = await stream2buffer(synth.AudioStream);

  const result:APIGatewayProxyResultV2 = {
    statusCode: 200,
    body: buffer.toString('base64'),
    isBase64Encoded: true,
    headers: {
      'Content-Type': 'audio/mpeg',
    },
  };

  return result;
};

async function stream2buffer(stream:Stream): Promise<Buffer> {
  return new Promise<Buffer>((resolve, reject) => {
    const _buf = Array<any>();
    stream.on('data', chunk => _buf.push(chunk));
    stream.on('end', () => resolve(Buffer.concat(_buf)));
    stream.on('error', err => reject(`error converting stream - ${err}`));
  });
}