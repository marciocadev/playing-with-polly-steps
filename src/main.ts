import { join } from 'path';
import { HttpApi, HttpMethod } from '@aws-cdk/aws-apigatewayv2-alpha';
import { HttpLambdaIntegration } from '@aws-cdk/aws-apigatewayv2-integrations-alpha';
import { App, Duration, Stack, StackProps } from 'aws-cdk-lib';
import { LoggingLevel, SlackChannelConfiguration } from 'aws-cdk-lib/aws-chatbot';
import { Alarm, ComparisonOperator } from 'aws-cdk-lib/aws-cloudwatch';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { Effect, ManagedPolicy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Architecture } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { Construct } from 'constructs';

export class MyStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    // especificamos o recurso que será permitido sintetizar
    // (no caso o Polly, diferente do Dynamo em que criamos
    // uma tabela o Polly temos de indicar que é todo o recurso com *)
    const pollyStatement = new PolicyStatement({
      effect: Effect.ALLOW,
      resources: ['*'],
      actions: ['polly:SynthesizeSpeech'],
    });

    const topic = new Topic(this, 'ErrorTopic', {
      topicName: 'polly-error-topic',
    });

    const lambda = new NodejsFunction(this, 'PollyLambda', {
      functionName: 'playing-with-polly',
      architecture: Architecture.ARM_64,
      entry: join(__dirname, 'lambda-fns/index.ts'),
      handler: 'handler',
      timeout: Duration.seconds(30),
    });
    lambda.addToRolePolicy(pollyStatement);
    const metric = lambda.metricErrors({
      period: Duration.minutes(1),
    });
    const alarm = new Alarm(this, 'PollyAlarm', {
      alarmName: 'polly-alarm-error',
      metric: metric,
      threshold: 1,
      evaluationPeriods: 1,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
    });
    const snsAction = new SnsAction(topic);
    alarm.addAlarmAction(snsAction);
    alarm.addOkAction(snsAction);
    // alarm.addInsufficientDataAction(snsAction);

    const rest = new HttpApi(this, 'PollyHttpApi', {
      apiName: 'polly-httpapi',
    });
    rest.addRoutes({
      path: '/speech/{voice}',
      methods: [HttpMethod.POST],
      integration: new HttpLambdaIntegration('Integration', lambda),
    });

    const slack = new SlackChannelConfiguration(this, 'MySlack', {
      slackChannelConfigurationName: 'error-message',
      slackWorkspaceId: 'T03L7TWLH09',
      slackChannelId: 'C03KEL2P2KH',
      notificationTopics: [topic],
      loggingLevel: LoggingLevel.INFO,
    });
    slack.role?.addManagedPolicy(
      ManagedPolicy.fromAwsManagedPolicyName('CloudWatchReadOnlyAccess'),
    );

  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new MyStack(app, 'playing-with-polly-steps-dev', { env: devEnv });
// new MyStack(app, 'playing-with-polly-steps-prod', { env: prodEnv });

app.synth();