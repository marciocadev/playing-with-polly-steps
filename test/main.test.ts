import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MyStack } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new MyStack(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});

describe('Lambda', () => {
  let app:App, stack: MyStack, template: Template;

  beforeAll(() => {
    // GIVEN
    app = new App();
    // WHEN
    stack = new MyStack(app, 'test');
    // THEN
    template = Template.fromStack(stack);
  });

  test('Lambda name', () => {
    template.hasResourceProperties('AWS::Lambda::Function', {
      FunctionName: 'playing-with-polly',
    });
  });

  test('Lambda arm architecture', () => {
    template.hasResourceProperties('AWS::Lambda::Function', {
      Architectures: ['arm64'],
    });
  });
});