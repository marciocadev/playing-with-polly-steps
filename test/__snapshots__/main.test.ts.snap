// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SlackChannelId": Object {
      "Description": "Enter slack channel id",
      "Type": "String",
    },
    "SlackWorkspaceId": Object {
      "Description": "Enter slack workspace id",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ErrorTopicA0904A23": Object {
      "Properties": Object {
        "TopicName": "polly-error-topic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "MySlackConfigurationRole8B3F1633": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "chatbot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchReadOnlyAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MySlackE351F6F6": Object {
      "Properties": Object {
        "ConfigurationName": "playing-with-polly",
        "IamRoleArn": Object {
          "Fn::GetAtt": Array [
            "MySlackConfigurationRole8B3F1633",
            "Arn",
          ],
        },
        "LoggingLevel": "INFO",
        "SlackChannelId": Object {
          "Ref": "SlackChannelId",
        },
        "SlackWorkspaceId": Object {
          "Ref": "SlackWorkspaceId",
        },
        "SnsTopicArns": Array [
          Object {
            "Ref": "ErrorTopicA0904A23",
          },
        ],
      },
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
    },
    "PollyAlarm440D1EF0": Object {
      "Properties": Object {
        "AlarmActions": Array [
          Object {
            "Ref": "ErrorTopicA0904A23",
          },
        ],
        "AlarmName": "polly-alarm-error",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "FunctionName",
            "Value": Object {
              "Ref": "PollyLambdaCC0E84BA",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "Errors",
        "Namespace": "AWS/Lambda",
        "OKActions": Array [
          Object {
            "Ref": "ErrorTopicA0904A23",
          },
        ],
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "PollyHttpApiD659CEE8": Object {
      "Properties": Object {
        "Name": "polly-httpapi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "PollyHttpApiDefaultStageEB762F2E": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "PollyHttpApiD659CEE8",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "PollyHttpApiPOSTspeechvoiceFCFCCD9F": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "PollyHttpApiD659CEE8",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "POST /speech/{voice}",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "PollyHttpApiPOSTspeechvoiceIntegration13C91CF5",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "PollyHttpApiPOSTspeechvoiceIntegration13C91CF5": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "PollyHttpApiD659CEE8",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "PollyLambdaCC0E84BA",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "PollyHttpApiPOSTspeechvoiceIntegrationPermissionFA1F35C8": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "PollyLambdaCC0E84BA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "PollyHttpApiD659CEE8",
              },
              "/*/*/speech/{voice}",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "PollyLambdaCC0E84BA": Object {
      "DependsOn": Array [
        "PollyLambdaServiceRoleDefaultPolicy81DBEB21",
        "PollyLambdaServiceRoleEEA63C4F",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "7ffcfe0bf89ebeddb6c0d0f4525effba13e920316917e6ca36989dce81baa35d.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "FunctionName": "playing-with-polly",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "PollyLambdaServiceRoleEEA63C4F",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "PollyLambdaServiceRoleDefaultPolicy81DBEB21": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "polly:SynthesizeSpeech",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PollyLambdaServiceRoleDefaultPolicy81DBEB21",
        "Roles": Array [
          Object {
            "Ref": "PollyLambdaServiceRoleEEA63C4F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PollyLambdaServiceRoleEEA63C4F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
